src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -3936,21 +3940,87 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
                 }
             }
 
-            // All options after this point require a ch local.
-            additionalDeclarations.Add("char ch;");
+            // Next, handle simple sets of two ranges where a simple bit manipulation transforms one into the other, e.g. [A-Za-z].
+            // This can be implemented as if it were a single range, with an additional bitwise operation.
+            if (!invariant &&
+                !RegexCharClass.IsNegated(charClass) &&
+                RegexCharClass.TryGetDoubleRange(charClass, out (char LowInclusive, char HighInclusive) rangeUnmasked, out (char LowInclusive, char HighInclusive) rangeMasked) &&
+                rangeUnmasked.LowInclusive != rangeUnmasked.HighInclusive && // more than a single char in range0 (just to simplify, since single chars should have already been handled)
+                rangeMasked.LowInclusive != rangeMasked.HighInclusive && // more than a single char in range1 (just to simplify, since single chars should have already been handled)
+                RegexCharClass.DifferByOneBit(rangeUnmasked.LowInclusive, rangeMasked.LowInclusive, out int lowMask) &&
+                RegexCharClass.DifferByOneBit(rangeUnmasked.HighInclusive, rangeMasked.HighInclusive, out int highMask) &&
+                lowMask == highMask)
+            {
+                // If the ranges are versions of each other that differ only by a single bit (e.g. cased ASCII letters), we can output a single comparison.
+                return $"((uint)(({chExpr} | 0x{lowMask:X}) - {Literal(rangeMasked.LowInclusive)}) {(negate ? ">" : "<=")} (uint)({Literal(rangeMasked.HighInclusive)} - {Literal(rangeMasked.LowInclusive)}))";
+            }
 
             // Analyze the character set more to determine what code to generate.
             RegexCharClass.CharClassAnalysisResults analysis = RegexCharClass.Analyze(charClass);
 
+            // Next, handle sets where the high - low + 1 range is <= 64.  In that case, we can emit
+            // a branchless lookup in a ulong that does not rely on loading any objects (e.g. the string-based
+            // lookup we use later).  This nicely handles common sets like [0-9A-Fa-f], [0-9a-f], [A-Za-z], etc.
+            if (!caseInsensitive &&
```

Comment:
Should this also check if !invariant?

Reply:
No, invariant is only relevant if caseInsensitive is true.